### PR TITLE
[media] Change local and remote audio write durations

### PR DESCRIPTION
**Pull request by osagie98**
This change sets the audio write ahead duration for local audio sinks to 0.5s, and the audio write ahead duration for remote audio sinks to 10s.

b/326652276

### CODE REVIEW

**Comment by osagie98** [comment]
Still need to test the idl changes.

**Comment by andrewsavage1** [comment]
In your commit message, please replace the gerrit link with the github link, especially with github being the source of truth: https://github.com/youtube/cobalt/pull/121

File: third_party/blink/renderer/core/html/media/html_media_element.idl
Comment (borongc) [inline third_party/blink/renderer/core/html/media/html_media_element.idl:100]: Consider to use Blink Runtime Enabled Feature to guard SbPlayer specific idl interfaces.

See https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/RuntimeEnabledFeatures.md;l=145?q=RuntimeEnabledFeatures%20-f:out&ss=chromium.

File: media/starboard/starboard_renderer.cc
Comment (borongc) [inline media/starboard/starboard_renderer.cc:700]: We should remove `COBALT_MEDIA_ENABLE_AUDIO_DURATION` completely, and since batched samples per write is not enabled as `max_audio_samples_per_write_=1`, we should let the code logic inside the block of `max_audio_samples_per_write_` to keep writing data up to `audio_write_duration_`.

File: third_party/blink/public/platform/web_media_player.h
Comment (borongc) [inline third_party/blink/public/platform/web_media_player.h:397]: Combine this `USE_STARBOARD_MEDIA` block with `GetSetBoundsCB()`, either move this up or move `GetSetBoundsCB` down.

File: third_party/blink/renderer/platform/media/web_media_player_impl.cc
Comment (borongc) [inline third_party/blink/renderer/platform/media/web_media_player_impl.cc:3975]: Same for here. Combine this `USE_STARBOARD_MEDIA` block with `WebMediaPlayerImpl::GetSetBoundsCB()`.

File: media/base/renderer_factory.cc
Comment (borongc) [inline media/base/renderer_factory.cc:13]: We shouldn't create a new function for `CreateRenderer()`. Instead, we should use `StarboardRendererFactory` to pass additional arguments needed by StarboardRenderer, and still use the same function of `CreateRenderer`. 

Refer https://github.com/youtube/cobalt/pull/4453 for how to pass `media_log_` to `StarboardRenderer` via `StarboardRendererFactory` without creating new `CreateRenderer()`.

File: media/starboard/starboard_renderer_factory.cc
Comment (borongc) [inline media/starboard/starboard_renderer_factory.cc:46]: We should avoid creating new functions for `CreateRenderer`, as all Chromium renderers use the same interface. This cannot be upstreamed if we do in this way.

File: content/renderer/media/media_factory.cc
Comment (borongc) [inline content/renderer/media/media_factory.cc:147]: Add a comment here for `kWriteDurationLocal` and `kWriteDurationRemote`.

File: media/base/renderer.h
Comment (borongc) [inline media/base/renderer.h:149]: Add a comment for `GetAudioConnectors()`.

File: media/filters/chunk_demuxer.cc
Comment (borongc) [inline media/filters/chunk_demuxer.cc:224]: nit: remove the new line.

File: media/filters/chunk_demuxer.cc
Comment (borongc) [inline media/filters/chunk_demuxer.cc:229]: nit: remove the new line.

File: media/starboard/starboard_renderer.cc
Comment (borongc) [reply media/starboard/starboard_renderer.cc:701]: No, just the macro of `COBALT_MEDIA_ENABLE_AUDIO_DURATION`.

File: third_party/blink/renderer/core/html/media/html_media_element.cc
Comment (borongc) [inline third_party/blink/renderer/core/html/media/html_media_element.cc:578]: We should try to enable this feature via command line in CobaltActivity.java (for ATV) via `--enable-blink-features`.

On Linux, feel free to put a TODO or put the corresponding place in the code.

File: third_party/blink/public/platform/web_source_buffer.h
Comment (borongc) [inline third_party/blink/public/platform/web_source_buffer.h:149]: nit: align the naming with other functions, like `WriteHead` in web_source_buffer.h.

Comment (borongc) [reply third_party/blink/public/platform/web_source_buffer.h:149]: And we should consider to allow this function to `=0`, so this is not necessary to be implemented.

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java
Comment (borongc) [inline cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:129]: @yell0wd0g For Cobalt-specific idl features, media team try to use blink runtime features for these attributes, please help review if this is the right way for us.

File: media/base/pipeline.h
Comment (borongc) [inline media/base/pipeline.h:271]: Could this default to `=0`, so once this is upstreamed, it is not required for other pipelines to implement it.

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java
Comment (osagie98) [reply cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:129]: Also, is it possible to do this on platforms that run the Cobalt binary directly, like Linux? We can add this param for a few platforms like Android, but I'm not sure if we'll need to run the Linux Cobalt binary with this param separately.

File: media/base/pipeline.h
Comment (borongc) [reply media/base/pipeline.h:271]: nit: Add comment for `GetAudioConnectors` as well.

File: third_party/blink/public/platform/web_source_buffer.h
Comment (borongc) [inline third_party/blink/public/platform/web_source_buffer.h:149]: nit: WriteHead

File: third_party/blink/renderer/core/html/media/html_media_element.cc
Comment (borongc) [inline third_party/blink/renderer/core/html/media/html_media_element.cc:1512]: nit: Add bug id.

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java
Comment (borongc) [reply cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:129]: For Linux, see https://github.com/youtube/cobalt/blob/main/cobalt/cobalt.cc.

**Comment by borongc** [comment]
Also, please update the commit message.

File: content/renderer/media/media_factory.cc
Comment (osagie98) [inline content/renderer/media/media_factory.cc:768]: Will create a bug for this once this PR is approved to merge.

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java
Comment (borongc) [inline cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java:129]: Wait @kaidokert and @yell0wd0g to confirm this is the right way to add new idl attributes in blink.

File: content/renderer/media/media_factory.cc
Comment (borongc) [inline content/renderer/media/media_factory.cc:152]: nit: consider to make these two numbers using `base::Milliseconds(500)` and `base::Seconds(10)`, so we don't need to add comments like `// 0.5 seconds` after them, as it is self-contained.

**Comment by borongc** [comment]
Waiting for @kaidokert and @yell0wd0g to confirm Cobalt specific idls using Blink runtime features. The rest of codes LGTM.

**Comment by osagie98** [comment]
@kaidokert or @yell0wd0g PTAL when free!

**Comment by andrewsavage1** [comment]
> @kaidokert or @yell0wd0g PTAL when free!

I'll defer but I thought that once https://github.com/youtube/cobalt/pull/4517 we'd be able to (and prefer to) use the javabridge for custom web apis.

**Comment by borongc** [comment]
https://github.com/youtube/cobalt/pull/4517 is only on Android, we may still think about how to do this on evergreen.

**Comment by andrewsavage1** [comment]
> #4517 is only on Android, we may still think about how to do this on evergreen.

It is only on Android, and that's the focus platform right now. We're going to implement web apis similarly to how Austin is currently implementing them once we move to a general solution, but we want separation between Cobalt and Blink code.

**Comment by osagie98** [comment]
I can go that way if it'll unblock this. Swapping to RuntimeEnabledFeatures later should be easy. Are we 100% certain that's the strategy for Android right now?

File: cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java
Comment (osagie98) [reply cobalt/android/apk/app/src/main/java/dev/cobalt/coat/CobaltActivity.java]: I'll address this in the writeHead PR, but I'll go ahead with a partial interface approach.

**Comment by osagie98** [comment]
I've cut down this PR to only address the audio write duration. I'll upload another PR to handle the SourceBuffer.writeHead changes. The h5vccaudioconnectors change will be in the background for now

**Review by borongc — APPROVED** [review]

File: content/renderer/media/media_factory.cc
Comment (niranjanyardi) [reply content/renderer/media/media_factory.cc:768]: Please use the TODO format here: cl/702469138

**Review by niranjanyardi — APPROVED** [review]
lg from a build perspective

**Review by andrewsavage1 — APPROVED** [review]

File: media/starboard/starboard_renderer.cc
Comment (xiaomings) [inline media/starboard/starboard_renderer.cc:596]: @osagie98 this hits DCHECK() on devel build when a video is playing to the end, due to calling timestamp() on an EOS buffer.

**Pull request merged**
